File: LinkedList.java
public class LinkedList<T> implements Iterable<T>{
    private int size = 0;
    private Node <T> head = null;
    private Node <T> tail = null;

    private class Node <T> {
        private T data;
        private Node<T> next, perv;

        public Node(T data, Node<T> next, Node<T> perv) {
            this.data = data;
            this.next = next;
            this.perv = perv;
        }

        @java.lang.Override
        public java.lang.String toString() {
            return data.toString();
        }
    }

    public void clear(){
        Node<T> t = head;
        while(t.next != null){
            Node<T> next = t.next;
            t.next = null;
            t.perv = null;
            t.data = null;
            t = null;
            t = next;
        }
        head = null;
        size = 0;
    }

    public int size(){
        return this.size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    public void append(T element){
        Node<T> newNode = new Node<>(element, null, tail);
        if(!isEmpty()){
            tail.next = newNode;
            tail = newNode;
        }
        else {
            head = newNode;
            tail = newNode;
        }
        size++;
    }

    public void insert(T element, int index){
        // Inserting to first
        if (index <= 0){
            Node<T> newNode = new Node<>(element, head, null);
            head.perv = newNode;
            head = newNode;
        }
        // Inserting to last
        else if(index >= size){
            Node<T> newNode = new Node<>(element, null, tail);
            tail.next = newNode;
            tail = newNode;
        }
        // Inserting to middle
        else {
            int trevIndex = 0;
            Node<T> pervNode;
            Node<T> nextNode = head;
            while(trevIndex < index){
                nextNode = nextNode.next;
                trevIndex++;
            }
            pervNode = nextNode.perv;
            Node<T> newNode = new Node<>(element, nextNode, pervNode);
            nextNode.perv = newNode;
            pervNode.next = newNode;
        }
        size++;
    }

    public T pop(){
        T returningData = tail.data;

        tail = tail.perv;
        size --;

        if(isEmpty()){
            head = null;
            tail = null;
        }
        else {
            tail.next = null;
        }

        return returningData;
    }

    public void remove(T element){
        Node<T> trav = head;

        while (trav != null){
            if (trav.data == element){
                if(trav.perv == null){ // removing head
                    head = trav.next;
                    trav.data = null;
                    trav.next = null;
                    trav.perv = null;
                    head.perv = null;
                }
                else if(trav.next == null) { // removing tail
                    tail = trav.perv;
                    trav.data = null;
                    trav.next = null;
                    trav.perv = null;
                    tail.next = null;
                }
                else { // removing from middle
                    trav.next.perv = trav.perv;
                    trav.perv.next = trav.next;
                    trav.data = null;
                    trav.next = null;
                    trav.perv = null;
                }
            }
            else {
                trav = trav.next;
            }
        }
    }

    public T get(int index){
        Node<T> trav = head;

        int travIndex = 0;
        while (trav != null){
            if(travIndex == index){
                return trav.data;
            }
            travIndex ++;
            trav = trav.next;
        }

        return null;
    }

    public int indexOf(T element){
        Node<T> trav = head;
        int index = 0;

        while (trav != null){
            if (trav.data == element){
                return index;
            }
            index++;
            trav = trav.next;
        }
        return -1;
    }

    public boolean contains(T element){
        Node<T> trav = head;

        while (trav != null){
            if (trav.data == element){
                return trav;
            }
            trav = trav.next;
        }
        return false;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[ ");
        Node<T> trav = head;
        while (trav != null) {
            sb.append(trav.data + ", ");
            trav = trav.next;
        }
        sb.append(" ]");
        return sb.toString();
    }

    @Override
    public java.util.Iterator<T> iterator() {
        return new java.util.Iterator<T>() {
            private Node<T> trav = head;

            @Override
            public boolean hasNext() {
                return trav != null;
            }

            @Override
            public T next() {
                T data = trav.data;
                trav = trav.next;
                return data;
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }
}

File: data-structures/linkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
    def __str__(self) -> str:
        return str(self.data)
        
class LinkedList:
    def __init__(self):
        self.head = None         
        self.size = 0   
            
    def insertAtBegin(self, data):
        self.size += 1
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        else:
            new_node.next = self.head
            self.head = new_node

    def insert(self, data, index):
        node = Node(data)
        
        current_node = self.head
        current_index = 0
        
        if current_index == index:
            self.insertAtBegin(data)
        else:
            while(current_node != None and current_index+1 != index):
                current_index = current_index+1
                current_node = current_node.next
 
            if current_node != None:
 
                node.next = current_node.next
                current_node.next = node
                self.size += 1
            else:
                raise IndexError("Index out of range")
        
    def append(self, data):
        self.size +=1
        node = Node(data)
        if self.head is None:
            self.head = node
            return
    
        current_node = self.head
        while current_node.next:
            current_node = current_node.next
    
        current_node.next = node

        
    def search(self, data):
        current = self.head
        while current != None and current.data != data:
            current = current.next
        return current
    
    def remove(self, index):
        if self.head == None:
            return
 
        current_node = self.head
        current_index = 0
        
        if current_index == index: # Removing the first element
            if self.head is None:
                    return
        
            current_node = self.head
            while(current_node.next.next):
                current_node = current_node.next
        
            current_node.next = None
            self.size -= 1
            
        else:
            while current_node != None and current_index+1 != index:
                current_index = current_index+1
                current_node = current_node.next
 
            if current_node != None:
                current_node.next = current_node.next.next
                self.size -= 1
            else:
                raise IndexError("Index out of range")
    
    def __str__(self) -> str:
        res = "["
        flag = True
        
        current = self.head
        
        while current:
            if flag:
                res += str(current.data)
                flag = False
            else:
                res += ", " + str(current.data)
            current = current.next
        return res + "]"
    
    
if __name__ == '__main__':
    import rand_arr_gen
    
    random_array = rand_arr_gen.random_list(10,0,50)
    
    linked_list = LinkedList()
    for i in random_array:
        linked_list.append(i)
        
    print("Linked List:", linked_list)
    
    item = random_array[5]
    print("Search result:", linked_list.search(item))
    